### PR TITLE
Feature/mf7

### PR DESCRIPTION
**Pull request by whaeck**
Full MF7 capability with MF8 MT457 as well ...

There is one outstanding issue with clang Debug indicating an 'undefined reference' in files/7.hpp that I have been unable to resolve. Other than that, all tests are green.

Coverage clocks in at 98%: [coverage-20190411-08:16.pdf](https://github.com/njoy/ENDFtk/files/3069099/coverage-20190411-08.16.pdf)

### CODE REVIEW

File: src/ENDFtk/section/7/2/src/print.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ *  @brief Print the MF7/MT2 section (includes SEND record)
+ *
+ *  @tparam OutputIterator   an output iterator
+ *
+ *  @param[in] it            the current position in the output
+ *  @param[in] MAT           the MAT number
+ *  @param[in] MF            the MF number
+ */
+template< typename OutputIterator >
+void print( OutputIterator& it, int MAT, int MF ) const {
```
Comment (jlconlin) [inline src/ENDFtk/section/7/2/src/print.hpp]: I don't think you need `MF` here as you know it is `7`. Perhaps you need it to maintain the same signature as other `print` function calls. If that's the case, then just leave off `MF` in the function signature and put `7` in it's place in the function body.

File: src/ENDFtk/section/7/4/ScatteringLawConstants.hpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+   */
+  double upperEnergyLimit() const { return ListRecord::list()[3]; }
+
+  /**
+   *  @brief Return the total free atom cross section for each scattering atom
+   *         type, stored in B(1), B(8) and B(14)
+   */
+  auto totalFreeCrossSections() const {
+
+    //! @todo this code no longer works since the ranges update, is there an
```
Comment (jlconlin) [inline src/ENDFtk/section/7/4/ScatteringLawConstants.hpp]: Let's have @apmccartney take a look at this.

File: src/ENDFtk/section/7/4/Tabulated/BetaValue.hpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ *  S(alpha,beta,T) data for a fixed value of beta from ENDF section MF7/MT4.
+ *  It contains a grid of alpha values with a corresponding S(T) grid for each
+ *  temperature as well as interpolation information as a function of alpha
+ *  and temperature.
+ *
+ *  Note: temperature interpolation of S(alpha,beta) is discouraged.
+ *
+ *  See ENDF102, section 7.4 for more information.
+ */
+class BetaValue {
```
Comment (jlconlin) [inline src/ENDFtk/section/7/4/Tabulated/BetaValue.hpp]: Why call this `BetaValue`?

File: src/ENDFtk/section/7/4/Tabulated/BetaValue/src/verifyBetaValues.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
+template< typename Array >
+static void
+verifyBetaValues( double beta, const Array& range ) {
+
+  auto iter = std::find_if( range.begin(), range.end(),
+                            [ beta ] ( const auto& v ) { return v != beta; } );
```
Comment (jlconlin) [inline src/ENDFtk/section/7/4/Tabulated/BetaValue/src/verifyBetaValues.hpp]: Does this just check that all the beta values are different than some `v`?

File: src/ENDFtk/section/8.hpp [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#include "ENDFtk/section/8/457.hpp"
```
Comment (jlconlin) [inline src/ENDFtk/section/8.hpp]: MF=8 stuff needs to be put in a different Pull Request.

File: src/ENDFtk/section/6/DiscreteTwoBodyScattering/LegendreCoefficients/src/ctor.hpp [hunk 6]
```diff
@@ -5,16 +5,11 @@ LegendreCoefficients( ListRecord&& list ) :
...
-                                          coefficients.size(),
-                                          std::move( coefficients ) ) ) {}
-  catch ( std::exception& e ) {
-
-    Log::info( "Encountered error while constructing Legendre coefficients "
-               "(LAW=2 LANG=0)" );
-    throw;
-  }
+LegendreCoefficients( double energy, std::vector< double >&& coefficients ) :
+  // this can never fail, try-catch would be unreachable
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/section/6/DiscreteTwoBodyScattering/LegendreCoefficients/src/ctor.hpp]: ListRecord takes the vector and sets the NPL value to the size of the vector. The basic constructor I use here for a ListRecord does not do any tests so it never fails, as a result you will never find anyone getting this to fail unless he changes the ListRecord constructor.

File: src/ENDFtk/file/7.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
+template<>
+class Type< 7 > {
+
+  /* MF7 has a limited number of possible sections (all optional) */
+  static constexpr auto optionalSections = hana::make_tuple( 2_c, 4_c );
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: to resolve the link time issue, define `optionalSections` as a static constexpr function.

```cpp
static constexpr auto optionalSections() {
  return hana::make_tuple( 2_c, 4_c );
}
```

File: src/ENDFtk/file/7.hpp [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  static constexpr auto sections = optionalSections;
+  static constexpr auto getSections() { return sections; }
+
+  /* convenience typedefs */
+  using Map =
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: 👍

File: src/ENDFtk/file/7.hpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  #include "ENDFtk/file/7/src/fill.hpp"
+
+public:
+
+  #include "ENDFtk/file/7/src/ctor.hpp"
+
+  #include "ENDFtk/file/7/src/sectionNumber.hpp"
+
+  template< typename Index >
+  decltype(auto) MT( Index sectionNo ) const {
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: use a crtp for this

File: src/ENDFtk/file/7.hpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    case 4: return bool( this->sectionMap[ 4_c ] );
+    default: return false;
+    }
+  }
+
+  bool
+  hasSectionNumber( int sectionNo ) const { return this->hasMT( sectionNo ); }
+
+  static constexpr auto MF(){ return 7; }
+  static constexpr auto fileNumber(){ return MF(); }
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: and this

Comment (whaeck) [reply src/ENDFtk/file/7.hpp]: To be honest, I was thinking about doing that :-)
It'll be useful for files as well that are in the same situation as MF7: MF1, MF2, MF8.

File: src/ENDFtk/file.hpp [hunk 11]
```diff
@@ -45,5 +45,6 @@ class Type {
 };
 
 #include "ENDFtk/file/1.hpp"
+#include "ENDFtk/file/7.hpp"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/file.hpp]: I'm glad you approve ;-)

File: src/ENDFtk/file/7.hpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
+template<>
+class Type< 7 > {
+
+  /* MF7 has a limited number of possible sections (all optional) */
+  static constexpr auto optionalSections = hana::make_tuple( 2_c, 4_c );
+  static constexpr auto sections = optionalSections;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/file/7.hpp]: As you know, most MF in the ENDF format allow for a single generic MT section. The exceptions to this rule are MF1, MF2, MF7 and MF8. These MF have special sections with specific MT numbers. Even then, these special cases have a few characteristics in common:
- some sections are required and must be present at all times (examples being MF1 MT451 and MF2 MT151)
- other sections are optional and do not need to be present if the specific MF is defined (e.g. all other MF1 sections, the special NJOY MF2 MT152 and MF153 or the MF7 MT2 and MT4)

As a result, it is better to define optional sections separately from the required sections. I would argue for putting this in a base class for all these cases but I haven't done that yet due to the link issue that will be resolved soon.

File: src/ENDFtk/file/7/src/ctor.hpp [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+template< typename BufferIterator >
+Type( StructureDivision& structureDivision,
+      BufferIterator& begin,
+      const BufferIterator& end,
+      long& lineNumber )
+try:
+  Type( this->read( decltype(sections){},
+                    structureDivision, begin, end, lineNumber,
+                    structureDivision.tail.MAT() ) ){
+      if ( not structureDivision.isFend() ){
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/file/7/src/ctor.hpp]: I don't think so, this is specific to the special case of MF7.

File: src/ENDFtk/section/6/ChargedParticleElasticScattering/LegendreCoefficients/test/LegendreCoefficients.test.cpp [hunk 14]
```diff
@@ -116,6 +102,22 @@ std::string chunk() {
     " 1.000000+0 2.000000+0 3.000000+0 4.000000+0                      9228 6  5     \n";
 }
 
+void verifyChunk( const LegendreCoefficients& chunk ) {
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/section/6/ChargedParticleElasticScattering/LegendreCoefficients/test/LegendreCoefficients.test.cpp]: Someone got on my case for not doing it this way, I forgot who that was ... ;-)

File: src/ENDFtk/section/6/ContinuumEnergyAngle/LegendreCoefficients/src/generateList.hpp [hunk 15]
```diff
@@ -32,6 +32,5 @@ generateList( unsigned int na,
            | ranges::view::transform(
```
Comment (apmccartney) [inline src/ENDFtk/section/6/ContinuumEnergyAngle/LegendreCoefficients/src/generateList.hpp]: The ranges adaptors are themselves callable objects. And as such, the lambdas are unnecessary

```cpp
namespace rv = ranges::view;

return rv::zip_with( ranges::view::concat, 
                     energies | rv::transform( rv::single ),
                     coefficients ) | ranges::view::join;
```

Bear in mind `concat` and `join` views are fairly expensive to iterate over. The `to_vector` call may have been warranted

File: src/ENDFtk/section/6/ContinuumEnergyAngle/src/checkLANG.hpp [hunk 16]
```diff
@@ -2,7 +2,7 @@ static void
 checkLANG( int lang ) {
 
   if ( ( lang != 1 ) && ( lang != 2 ) && ( lang != 11 ) &&
```
Comment (apmccartney) [inline src/ENDFtk/section/6/ContinuumEnergyAngle/src/checkLANG.hpp]: I prefer `and` to `&&` myself, but not a big deal

File: src/ENDFtk/section/7/2.hpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+  /* constructor */
+  #include "ENDFtk/section/7/2/src/ctor.hpp"
+
+  /* get methods */
+
+  /**
+   *  @brief Return the LTHR flag (coherent or incoherent elastic scattering)
+   */
+  int LTHR() const { return std::visit( [&] ( const auto& v ) -> long
```
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/section/7/2.hpp]: The capture is uneccessary

File: src/ENDFtk/section/7/2/CoherentElastic/src/generateTemperatures.hpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
+static std::vector< ListRecord >
+generateTemperatures( std::vector< double >&& temperatures,
+                      std::vector< long >&& li,
+                      std::vector< std::vector< double > >&& s ) {
+  unsigned int size = temperatures.size() - 1;
+  std::vector< ListRecord > sequence;
```
[the remark this answers is not in the document]
Comment (apmccartney) [reply src/ENDFtk/section/7/2/CoherentElastic/src/generateTemperatures.hpp]: return value optimization, but you have the right instinct

File: src/ENDFtk/section/7/2/CoherentElastic.hpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+ *  The S(E) functions are given as tabulated S(E) for a set of temperature
+ *  values.
+ *
+ *  See ENDF102, section 7.2 for more information.
+ */
+class CoherentElastic {
+
+  /* members */
+  TabulationRecord principal_;
+  std::vector< ListRecord > temperatures_;
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/section/7/2/CoherentElastic.hpp]: principal_ contains S(E,T0) for the first temperature. If more than one temperature is defined, each additional temperature has a ListRecord with the corrsponding S(E,Ti) values.

Now, it would appear that I do not check that NP in the principal_ TAB1 is equal to NPL in each ListRecord.

File: src/ENDFtk/section/7/4/Tabulated/BetaValue/src/verifyBetaValues.hpp (hunk 4 above)
Comment (whaeck) [reply src/ENDFtk/section/7/4/Tabulated/BetaValue/src/verifyBetaValues.hpp]: The beta value appears on the initial TAB1 but also the following LISTs (if more than 1 temperature is defined). This checks that all beta values in a single BetaValue are the same.

File: src/ENDFtk/section/7/4.hpp [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+  /* fields */
+  int lat_;
+  int lasym_;
+
+  ScatteringLawConstants b_;
+  ScatteringLaw law_;
+
+  EffectiveTemperature principal_;
+  std::vector< std::optional< EffectiveTemperature > > secondary_;
```
Comment (apmccartney) [inline src/ENDFtk/section/7/4.hpp]: let's talk about this field

File: src/ENDFtk/section/7/4/ScatteringLawConstants.hpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    //!       alternative as the current operational code is more of a hack?
+    // return ranges::view::concat(
+    //          ranges::view::single( ListRecord::list()[0] ), 
+    //          ranges::view::drop_exactly( ListRecord::list(), 7 )
+    //                | ranges::view::stride( 6 )  );
+
```
Comment (apmccartney) [inline src/ENDFtk/section/7/4/ScatteringLawConstants.hpp]: David's been bugging them about various concat bugs, most of which have been resolved. Might be worth another update.

File: src/ENDFtk/section/6/ContinuumEnergyAngle/LegendreCoefficients/src/generateList.hpp (hunk 15 above)
Comment (whaeck) [reply src/ENDFtk/section/6/ContinuumEnergyAngle/LegendreCoefficients/src/generateList.hpp]: That'll make it a bit more readable, thanks :-)

The function actually returns a vector so the range will be put into a vector anyway before being given to the ListRecord constructor. I would assume to_vector is not required in this case.

File: src/ENDFtk/section/7/4/ScatteringLawConstants.hpp [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    if ( ListRecord::NPL() > 13 ) { result.push_back( ListRecord::list()[13] ); }
+    return result;
+  }
+
+  /**
+   *  @brief Return the ratio of the atomic weight to the neutron mass for each
+   *         scattering atom type, stored in B(3), B(9) and B(15)
+   */
+  auto atomicWeightRatios() const {
+    return ranges::view::drop_exactly( ListRecord::list(), 2 )
```
Comment (apmccartney) [inline src/ENDFtk/section/7/4/ScatteringLawConstants.hpp]: Certainly nothing wrong with using it as a function this way, but to make sure you're aware, `drop_exactly` is pipeable as well

```cpp
return 
  ListRecord::list()
  | ranges::view::drop_exactly(2)
  | ranges::view::stride(6);
```

Comment (whaeck) [reply src/ENDFtk/section/7/4/ScatteringLawConstants.hpp]: Damn, my ranges lingo is rusty :-)

File: src/ENDFtk/section/7/4.hpp (hunk 20 above)
Comment (whaeck) [reply src/ENDFtk/section/7/4.hpp]: Yeah, this is a special one. An incoherent inelastic scattering section (MF7 MT4) defines a principal scatterer which has its EffectiveTemperature associated to it (it must be given). On the other hand, the section may also define a number of secondary scatterers (generally speaking I think the number of secondary scatterers is between 0 and 3). Only if the analytical function type is set to 0 for a given secondary scatterer (set in the ScatteringLawConstants) will there be an EffectiveTemperature given, hence the optional EffectiveTemperature.

The size of the vector is equal to the number of secondary scatterers defined in the section.

File: src/ENDFtk/section/6/ContinuumEnergyAngle.hpp [hunk 23]
```diff
@@ -11,8 +11,6 @@ class ContinuumEnergyAngle {
   InterpolationSequenceRecord< SubSection > data_;
 
   /* auxiliary functions */
-  #include "ENDFtk/section/6/ContinuumEnergyAngle/src/checkLANG.hpp"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/section/6/ContinuumEnergyAngle.hpp]: The read functions take care of verifying the LANG value while it is impossible to construct a ContinuumEnergyAngle with the wrong LANG (lang has been eliminated from the input arguments and is now taken from the sequence). All elements of the sequence still need the same LANG, which is now also tested for.

File: src/ENDFtk/section/7/4/Tabulated/BetaValue.hpp (hunk 3 above)
Comment (whaeck) [reply src/ENDFtk/section/7/4/Tabulated/BetaValue.hpp]: Renamed to ScatteringFunction.

File: src/ENDFtk/section/7/4/ScatteringLawConstants/src/verifySize.hpp [hunk 24]
```diff
@@ -0,0 +1,12 @@
+static void
+verifySize( int NI, int NS ) {
+
+  if ( NI != 6 * ( NS + 1 ) ) {
```
Comment (whaeck) [inline src/ENDFtk/section/7/4/ScatteringLawConstants/src/verifySize.hpp]: verify that NS <= 3

**Comment by apmccartney** [comment]
reference for tomorrows discussion

```c++
  template<typename Index>                                                                                                                                    
  using Required_t =                                                                                                                                          
    hana::pair<Index, section::Type<Derived::MF(), Index::value>>;                                                                                             
                                                                                                                                                              
  template<typename Index>                                                                                                                                    
  using Optional_t =                                                                                                                                          
    hana::pair                                                                                                                                                
    <Index, std::optional<section::Type<Derived::MF(), Index::value>>>;                                                                                        
                                                                                                                                                              
  template<typename... Pairs>                                                                                                                                 
  using Map_t = decltype(hana::make_map(std::declval<Pairs>()...));                                                                                           
                                                                                                                                                              
  static constexpr auto sections(){                                                                                                                           
    return hana::sort(hana::concat(Derived::requiredSections(),                                                                                               
                                   Derived::optionalSections()));                                                                                             
  }                                                                                                                                                           
                                                                                                                                                              
  static constexpr auto derivedMap(){                                                                                                                            
    const auto isRequired =                                                                                                                                         
      hana::partial(hana::in, Derived::requiredSections());                                                                                                   
    const auto makeRequired =                                                                                                                                       
      hana::compose(hana::template_<Required_t>, hana::make_type);                                                                                            
    const auto makeOptional =                                                                                                                                       
      hana::compose(hana::template_<Optional_t>, hana::make_type);                                                                                            
    const auto makePair =                                                                                                                                           
      hana::demux(hana::if_)(isRequired, makeRequired, makeOptional);                                                                                         
    const auto makeMap = hana::template_<Map_t>;                                                                                                                    
                                                                                                                                                              
    return hana::unpack(sections(), hana::on(makeMap, makePair));                                                                                    
  }    
```

File: src/ENDFtk/file/1.hpp [hunk 25]
```diff
@@ -1,9 +1,15 @@
...
-class Type< 1 > {
-protected:
-  /* MF1 has a limited number of possible sections */
-  static constexpr auto optionalSections =
-    hana::make_tuple( 452_c, 455_c, 456_c, 458_c, 460_c );
+class Type< 1 > : public Base< Type< 1 > >,
+                  public SpecialBase< Type< 1 > > {
+
+  friend SpecialBase< Type< 1 > >;
+
```
Comment (apmccartney) [inline src/ENDFtk/file/1.hpp]: please add `requireSections` static constexpr method

File: src/ENDFtk/file/1/src/ctor.hpp [hunk 26]
```diff
@@ -1,160 +1,36 @@
...
-}
-*/
-
-// stopgap until remainder of file 1 complete
-public:
-Type( section::Type< 1, 451 >&& section451 ) :
-  sectionMap( hana::make_map
-              ( hana::make_pair( 451_c, std::move(section451) ) ) ){}
+template< typename... Args >
+Type( section::Type< 1, 451 >&& section_, Args&&... args ) :
```
Comment (apmccartney) [inline src/ENDFtk/file/1/src/ctor.hpp]: We don't actually care what the MT of the first section argument is.
```c++
template< long long MT, typename... Args >
Type( StructureDivision& structureDivision,
Type( section::Type< 1,  MT >&& section_, Args&&... args )
```

File: src/ENDFtk/file/1/src/ctor.hpp [hunk 27]
```diff
@@ -1,160 +1,36 @@
...
       const BufferIterator& end,
       long& lineNumber )
 try:
-  Type( this->read( 451_c, structureDivision, begin, end, lineNumber ) ){
+  Type( this->read( 451_c, structureDivision, begin, end, lineNumber ) ) {
     if ( not structureDivision.isFend() ){
-      if ( structureDivision.isHead() ){
+      if ( structureDivision.isSend() ){
+        Log::error("Encountered duplicate SEND record in File 1");
+        Log::info("Section number: {}", asHead( structureDivision ).section() );
```
Comment (apmccartney) [inline src/ENDFtk/file/1/src/ctor.hpp]: it's been awhile. Isn't the MT on a send record always zero?

File: src/ENDFtk/file/1/src/fill.hpp [hunk 28]
```diff
@@ -1,21 +1,35 @@
 template< typename... Sections >
 static Map
-fill( section::Type<1, 451> >&& section451, Sections&&... args ) const {
+fill( section::Type< 1, 451 >&& section451, Sections&&... sections ) {
```
Comment (apmccartney) [inline src/ENDFtk/file/1/src/fill.hpp]: Let's do this lift into `File/Base.hpp` together today

File: src/ENDFtk/file/7.hpp [hunk 29]
```diff
@@ -0,0 +1,45 @@
+template<>
+class Type< 7 > : public Base< Type< 7 > >,
+                  public SpecialBase< Type< 7 > > {
+
+  friend SpecialBase< Type< 7 > >;
+
+  /* MF7 has a limited number of possible sections (all optional) */
+  static constexpr auto optionalSections() { return hana::make_tuple( 2_c, 4_c ); }
+  static constexpr auto sections() { return optionalSections(); }
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: let's provide this method in the File::Base

File: src/ENDFtk/file/7.hpp [hunk 30]
```diff
@@ -0,0 +1,45 @@
+template<>
+class Type< 7 > : public Base< Type< 7 > >,
+                  public SpecialBase< Type< 7 > > {
+
+  friend SpecialBase< Type< 7 > >;
+
+  /* MF7 has a limited number of possible sections (all optional) */
+  static constexpr auto optionalSections() { return hana::make_tuple( 2_c, 4_c ); }
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: please add a `requiredSections` static constexpr method

File: src/ENDFtk/file/7.hpp [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+  static constexpr auto sections() { return optionalSections(); }
+
+  /* convenience typedefs */
+  using Map =
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: To be provided by File::Base

File: src/ENDFtk/file/7.hpp [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+  /* auxiliary functions */
+  #include "ENDFtk/file/7/src/read.hpp"
+  #include "ENDFtk/file/7/src/fill.hpp"
+
+public:
+
+  #include "ENDFtk/file/7/src/ctor.hpp"
+
+  bool
+  hasSection( int sectionNo ) const {
```
Comment (apmccartney) [inline src/ENDFtk/file/7.hpp]: We can provide this in `File::Base` with a little effort

File: src/ENDFtk/file/7/src/fill.hpp [hunk 33]
```diff
@@ -0,0 +1,33 @@
+template< int sectionNo, typename... Sections >
+static Map
```
Comment (apmccartney) [inline src/ENDFtk/file/7/src/fill.hpp]: repeated code to be lifted

File: src/ENDFtk/file/7/src/print.hpp [hunk 34]
```diff
@@ -0,0 +1,10 @@
+template< typename OutputIterator >
```
Comment (apmccartney) [inline src/ENDFtk/file/7/src/print.hpp]: repeated code to be lifted

File: src/ENDFtk/file/7/src/read.hpp [hunk 35]
```diff
@@ -0,0 +1,66 @@
+template< long long SectionNo, typename BufferIterator >
```
Comment (apmccartney) [inline src/ENDFtk/file/7/src/read.hpp]: repeated code to be lifted

File: src/ENDFtk/file/8.hpp [hunk 36]
```diff
@@ -0,0 +1,50 @@
+template<>
+class Type< 8 > : public Base< Type< 8 > >,
+                  public SpecialBase< Type< 8 > > {
+
+  friend SpecialBase< Type< 8 > >;
+
+  /* MF8 has enumerated sections */
+  static constexpr auto optionalSections() { return hana::make_tuple( /*454_c,*/ 457_c/*, 457_c*/ ); }
```
Comment (apmccartney) [inline src/ENDFtk/file/8.hpp]: please add a requiredSections constexpr static method

File: src/ENDFtk/file/8.hpp [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+  static constexpr auto sections() { return optionalSections(); }
```
Comment (apmccartney) [inline src/ENDFtk/file/8.hpp]: repeated code to be lifted

File: src/ENDFtk/file/8.hpp [hunk 38]
```diff
@@ -0,0 +1,50 @@
...
+
+  /* convenience typedefs */
+  using Map =
```
Comment (apmccartney) [inline src/ENDFtk/file/8.hpp]: WIll be handled by Base

File: src/ENDFtk/file/8/src/ctor.hpp [hunk 39]
```diff
@@ -0,0 +1,42 @@
+protected:
+Type( Map&& map ) : sectionMap( std::move(map) ){}
+
+public:
+template< int sectionNo, typename... Args >
+Type( section::Type< 8, sectionNo >&& section_, Args&&... args ) :
```
Comment (apmccartney) [inline src/ENDFtk/file/8/src/ctor.hpp]: redundant code to be lifted

File: src/ENDFtk/file/8/src/ctor.hpp [hunk 40]
```diff
@@ -0,0 +1,42 @@
...
+  Type( fill( std::move( section_ ), std::forward< Args >( args )... ) ) {}
+
+template< typename BufferIterator >
+Type( StructureDivision& structureDivision,
```
Comment (apmccartney) [inline src/ENDFtk/file/8/src/ctor.hpp]: redundant code to be lifted

File: src/ENDFtk/file/8/src/fill.hpp [hunk 41]
```diff
@@ -0,0 +1,33 @@
+template< int sectionNo, typename... Sections >
+static Map
+fill( section::Type< 8, sectionNo >&& section, Sections&&... sections ) {
```
Comment (apmccartney) [inline src/ENDFtk/file/8/src/fill.hpp]: redundant code to be lifted

File: src/ENDFtk/file/8/src/print.hpp [hunk 42]
```diff
@@ -0,0 +1,10 @@
+template< typename OutputIterator >
```
Comment (apmccartney) [inline src/ENDFtk/file/8/src/print.hpp]: redundant code to be lifted

File: src/ENDFtk/file/8/src/read.hpp [hunk 43]
```diff
@@ -0,0 +1,66 @@
+template< long long SectionNo, typename BufferIterator >
```
Comment (apmccartney) [inline src/ENDFtk/file/8/src/read.hpp]: redundant code to be lifted

File: src/ENDFtk/file/Type.hpp [hunk 44]
```diff
@@ -0,0 +1,40 @@
...
+  /* convenience typedefs */
+  using Section = section::Type< FileNumber >;
+  using iterator = typename std::vector< Section >::iterator;
+  using const_iterator = typename std::vector< Section >::const_iterator;
+
+protected :
+
+  /* fields */
+  std::vector< Section > sectionVector;
+  tsl::hopscotch_map< int, const Section* > sectionMap;
```
Comment (apmccartney) [inline src/ENDFtk/file/Type.hpp]: Let's go ahead and make the jump to the map

File: src/ENDFtk/section/6/ChargedParticleElasticScattering/NuclearAmplitudeExpansion/src/generateList.hpp [hunk 45]
```diff
@@ -18,7 +18,7 @@ generateList( unsigned int NL,
     Log::info( "ai.size(): {}", ai.size() );
     throw std::exception();
   }
-  if ( ! ( ( b.size() == NL + 1 ) || ( b.size() == 2 * NL + 1 ) ) ) {
+  if ( ! ( ( b.size() == NL + 1 ) or ( b.size() == 2 * NL + 1 ) ) ) {
```
Comment (apmccartney) [inline src/ENDFtk/section/6/ChargedParticleElasticScattering/NuclearAmplitudeExpansion/src/generateList.hpp]: if we're touching these anyway, `not` preferable to `!`

File: src/ENDFtk/section/7/2/src/print.hpp (hunk 1 above)
Comment (apmccartney) [reply src/ENDFtk/section/7/2/src/print.hpp]: As section types are file-specific, I don't think this is ever necessary.

that said, let's defer this to another pull

File: src/ENDFtk/section/7/2/CoherentElastic/src/verifyTemperatures.hpp [hunk 46]
```diff
@@ -0,0 +1,20 @@
...
+static void
+verifyTemperatures( int NP, const Array& temperatures ) {
+
+  // lambdas for verifying
+  auto verifySize = [NP] ( const auto& value ) -> bool
+                         { return value.NPL() == NP; };
+  auto sum = [] ( bool left, bool right ) -> bool
+                { return left and right; };
+
+  if ( ranges::accumulate(
```
Comment (apmccartney) [inline src/ENDFtk/section/7/2/CoherentElastic/src/verifyTemperatures.hpp]: short circuiting with `ranges::find_if_not` would be preferred

File: src/ENDFtk/section/7/2/src/NC.hpp [hunk 47]
```diff
@@ -0,0 +1,11 @@
+/** 
+ *  @brief Return the number of lines in this MF7/MT2 section
+ */
+long NC() const {
+  long NC = 1;
```
Comment (apmccartney) [inline src/ENDFtk/section/7/2/src/NC.hpp]: not worth changing, but prefer to minimize mutation

```c++
return 1 + std::visit( [&] ( const auto& v ){ return v.NC(); }, this->law_ );
```

File: src/ENDFtk/section/7/4/ScatteringLawConstants/src/checkLLN.hpp [hunk 48]
```diff
@@ -0,0 +1,12 @@
+static void
+checkLLN( int value ) {
+
+  if ( ( value != 0 ) and ( value != 1 ) ) {
+
+    Log::error( "Encountered illegal LLN value" );
```
Comment (apmccartney) [inline src/ENDFtk/section/7/4/ScatteringLawConstants/src/checkLLN.hpp]: by convention, info preceeds error messages, no?

File: src/ENDFtk/section/7/4/Tabulated/ScatteringFunction/src/verifyBetaValues.hpp [hunk 49]
```diff
@@ -0,0 +1,15 @@
+template< typename Array >
+static void
+verifyBetaValues( double beta, const Array& range ) {
+
+  auto iter = std::find_if( range.begin(), range.end(),
+                            [ beta ] ( const auto& v ) { return v != beta; } );
```
Comment (apmccartney) [inline src/ENDFtk/section/7/4/Tabulated/ScatteringFunction/src/verifyBetaValues.hpp]: part of the reason the compile times are large-ish is because we generate a lot of intermediate types. Consider preferring types with more reuse, e.g.

```c++
auto iter = std::find_if( range.begin(), range.end(), hana::not_equal.to(beta));
```

File: src/ENDFtk/section/7/4/ScatteringLawConstants/src/checkLLN.hpp (hunk 48 above)
Comment (jlconlin) [reply src/ENDFtk/section/7/4/ScatteringLawConstants/src/checkLLN.hpp]: No, info messages have always come after the error messages.

File: src/ENDFtk/file/1/src/ctor.hpp (hunk 27 above)
Comment (whaeck) [reply src/ENDFtk/file/1/src/ctor.hpp]: Yep, that's a copy paste followed by forgetting to actually read the code :-)

File: src/ENDFtk/section/7/2/CoherentElastic.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (whaeck) [reply src/ENDFtk/section/7/2/CoherentElastic.hpp]: Maybe rename temperature_ to something more appropriate (like sOfTemp_). Think about it.

File: src/ENDFtk/section/7/2/CoherentElastic/src/verifyTemperatures.hpp (hunk 46 above)
Comment (whaeck) [reply src/ENDFtk/section/7/2/CoherentElastic/src/verifyTemperatures.hpp]: Done, in several locations

File: src/ENDFtk/file/Type.hpp (hunk 44 above)
Comment (whaeck) [reply src/ENDFtk/file/Type.hpp]: I changed this to std::map, you can have a look.

**Comment by jlconlin** [comment]
I really like the `verifyEND` methods. I think it cleans the code up quite a bit. It also probably improves the coverage!
